### PR TITLE
Feat/3433 prep randomness sampling to current spec: ticket iterator

### PR DESCRIPTION
**Pull request by ZenGround0**
This PR updates randomness sampling to make use of all ticket array values per the spec.  Closes out #3433, don't review until #2223 is merged.

This is a protocol breaking change.  If we can get it in for 0.5.x awesome, if not we need to make an upgrade.

### CODE REVIEW

File: chain/traversal.go [hunk 1]
```diff
@@ -53,6 +53,63 @@ func (it *TipsetIterator) Next() error {
...
+func (it *TicketIterator) Complete() bool {
+	return it.tips.Complete()
+}
+
+// Next advances the iterator to the next ticket. Tickets are taken from the
+// block with the minimum final ticket in its ticket array
+// https://github.com/filecoin-project/specs/blob/master/definitions.md#ticket-chain
+func (it *TicketIterator) Next() error {
+	if it.tips.Complete() {
+		return nil
```
Comment (anorth) [inline chain/traversal.go]: Probably a bug so return an error.

File: chain/traversal.go [hunk 2]
```diff
@@ -53,6 +53,63 @@ func (it *TipsetIterator) Next() error {
...
+	}
+
+	// Advance internal tipset iterator
+	if it.idx == 0 {
+		if err := it.tips.Next(); err != nil {
```
Comment (anorth) [inline chain/traversal.go]: After `Next` you must check `Complete`.

File: chain/traversal.go [hunk 3]
```diff
@@ -53,6 +53,63 @@ func (it *TipsetIterator) Next() error {
...
+			return err
+		}
+		it.idx = len(it.tips.Value().At(0).Tickets) - 1
+		if it.tips.Value().Equals(types.UndefTipSet) {
```
Comment (anorth) [inline chain/traversal.go]: The line above will panic if this comparison is true.

**Comment by ZenGround0** [comment]
This is low priority to merge because it won't make it into the node's live code path until we launch an associated protocol change.  It is also possible that ticket sampling is going to change significantly in the spec between now and then in which case we'll get rid of this anyway.

I'm not including the integration step here because that should be done separately with a protocol breaking change.  I will file a quick follow on draft PR demonstrating how that will look.

**Comment by anorth** [comment]
@zen if the PR as is constitutes a protocol change, please close it for now. Please do put up any parts of it that we can land without a protocol upgrade.

**Pull request closed without merging**